### PR TITLE
Add WhatsApp to NativeApp

### PR DESCRIPTION
**Pull request by dhakalaashish**
This PR adds WhatsApp to nativeapp and updates the server implementation as well. It also upgrades react-native-auth0 from v2 to v3.

### CODE REVIEW

File: apps/nativeapp/app/components/phoneInput/PhoneInput.tsx [hunk 1]
```diff
@@ -4,7 +4,7 @@ import PhoneNoInput from 'react-native-phone-number-input';
 import {CountryCode} from 'react-native-country-picker-modal';
 
 import {Colors, Typography} from '../../styles';
-import {EXCLUDED_COUNTRIES} from '../../constants';
+import {EXCLUDED_COUNTRIES, EXCLUDED_COUNTRIES_WHATSAPP} from '../../constants';
```
Comment (mohitb35) [inline apps/nativeapp/app/components/phoneInput/PhoneInput.tsx:7]: I'd suggest renaming EXCLUDED_COUNTRIES to indicate it includes countries where SMS is disabled.

File: apps/server/src/server/api/routers/alertMethod.ts [hunk 2]
```diff
@@ -218,11 +218,22 @@ export const alertMethodRouter = createTRPCRouter({
...
                         code: 'BAD_REQUEST',
-                        message: `Destination is restricted due to country limitations`,
+                        message: `Phone number is restricted due to country limitations.`,
+                    });
+                }
+            }
+            // If Method is whatsapp, restrict phone number to only allowed countries
+            if(input.method === 'whatsapp'){
+                // Check if the destination falls inside of accepted countries
+                const isDestinationAccepted = !isPhoneNumberRestricted('whatsapp', input.destination);
```
Comment (mohitb35) [inline apps/server/src/server/api/routers/alertMethod.ts:232]: Is this check not needed for `verify:`?

File: apps/nativeapp/app/screens/Verification/Verification.tsx [hunk 3]
```diff
@@ -84,9 +84,21 @@ const Verification = ({navigation, route}) => {
...
-    if (verificationType === 'Sms' && limitDestination) {
+    if (
+      (verificationType === 'Sms' && limitDestination)
+    ) {
+      return toast.show(
+        `Apologies, ${verificationType === 'Sms' ? 'sms' : 'WhatsApp'} services to your region are currently unsupported. Please select a different method.`,
+        {
+          type: 'warning',
+        },
+      );
```
Comment (mohitb35) [inline apps/nativeapp/app/screens/Verification/Verification.tsx:95]: There's a double check for `verificationType === 'Sms` here - that can be removed

File: apps/nativeapp/app/screens/Settings/Settings.tsx [hunk 4]
```diff
@@ -449,27 +450,27 @@ const Settings = ({navigation}) => {
...
+      detectionMethods = [...detectionMethods, 'GEOSTATIONARY'];
+    }
+    updateUser.mutate({
+      json: {
+        body: {
+          detectionMethods,
+        },
+      },
+    });
+  };
```
Comment (mohitb35) [inline apps/nativeapp/app/screens/Settings/Settings.tsx:473]: Needs to be commented

File: apps/nativeapp/app/screens/Login/Login.tsx [hunk 5]
```diff
@@ -46,7 +46,7 @@ const Login = () => {
           prompt: 'login',
           audience: 'urn:plant-for-the-planet',
         },
-        {ephemeralSession: false},
+        {ephemeralSession: false, useLegacyCallbackUrl: true},
```
Comment (mohitb35) [inline apps/nativeapp/app/screens/Login/Login.tsx:49]: Is this an intermediate step? What stops us from updating the Auth0 callback Url, or adding a new one?

**Review by mohitb35 — CHANGES_REQUESTED** [review]
@dhakalaashish I've requested some minor changes. And raised some points of clarification.

File: apps/server/src/server/api/routers/alertMethod.ts (hunk 2 above)
Comment (dhakalaashish) [reply apps/server/src/server/api/routers/alertMethod.ts:232]: We are dynamically restricting phone numbers conditionally. Only Indian numbers are restricted for WhatsApp, while sms are restricted for about 8 countries.

File: apps/nativeapp/app/screens/Login/Login.tsx (hunk 5 above)
Comment (dhakalaashish) [reply apps/nativeapp/app/screens/Login/Login.tsx:49]: This Boolean is necessary to ensure conflict-free migration from react-native-auth0 from v2 to v3.

File: apps/nativeapp/app/screens/Settings/Settings.tsx (hunk 4 above)
Comment (dhakalaashish) [reply apps/nativeapp/app/screens/Settings/Settings.tsx:473]: Done. Thank you.

File: apps/nativeapp/app/screens/Verification/Verification.tsx (hunk 3 above)
Comment (dhakalaashish) [reply apps/nativeapp/app/screens/Verification/Verification.tsx:95]: Done. Thank you.

**Comment by dhakalaashish** [comment]
@mohitb35 Thank you for your feedback. I've made most of the suggested changes, but retained some for correctness.

**Review by mohitb35 — APPROVED** [review]
LGTM once conflicts are resolved

**Pull request merged**